packages/flutter/lib/src/material/app.dart

```diff
@@ -993,29 +993,68 @@ class _MaterialAppState extends State<MaterialApp> {
   }
 
   ThemeData _themeBuilder(BuildContext context) {
-    ThemeData? theme;
-    // Resolve which theme to use based on brightness and high contrast.
     final ThemeMode mode = widget.themeMode ?? ThemeMode.system;
     final Brightness platformBrightness = MediaQuery.platformBrightnessOf(context);
     final bool useDarkTheme =
         mode == ThemeMode.dark ||
         (mode == ThemeMode.system && platformBrightness == ui.Brightness.dark);
     final bool highContrast = MediaQuery.highContrastOf(context);
-    if (useDarkTheme && highContrast && widget.highContrastDarkTheme != null) {
-      theme = widget.highContrastDarkTheme;
-    } else if (useDarkTheme && widget.darkTheme != null) {
-      theme = widget.darkTheme;
-    } else if (highContrast && widget.highContrastTheme != null) {
-      theme = widget.highContrastTheme;
-    }
-    theme ??= widget.theme ?? ThemeData.light();
+
+    final ThemeData theme = _buildThemeForMode(
+      useDarkTheme: useDarkTheme,
+      highContrast: highContrast,
+    );
+
     SystemChrome.setSystemUIOverlayStyle(
       theme.brightness == Brightness.dark ? SystemUiOverlayStyle.light : SystemUiOverlayStyle.dark,
     );
 
     return theme;
   }
 
+  ThemeData _buildThemeForMode({required bool useDarkTheme, required bool highContrast}) {
+    // First step is to pick the appropriate theme property based on this matrix:
+    //
+    // +-------------------+--------------------------+------------------------------+
+    // |                   | Light                    | Dark                         |
+    // +-------------------+--------------------------+------------------------------+
+    // | High contrast OFF | widget.theme             | widget.darkTheme             |
+    // |                   |                          |                              |
+    // | High contrast ON  | widget.highContrastTheme | widget.highContrastDarkTheme |
+    // +-------------------+--------------------------+------------------------------+
+    //
+    // When a high contrast theme is unavailable, the non-high-contrast theme for the same
+    // brightness is used. Example, if `highContrastDarkTheme` is null, `darkTheme` is used.
+
+    ThemeData? theme = switch ((useDarkTheme, highContrast)) {
+      ((false, false)) => widget.theme,
+      ((false, true)) => widget.highContrastTheme ?? widget.theme,
```

Comment:
do we need to fallback here if we doing the fallback in line 1042?

To make the code a bit more consistent
Either doing all fallback here and remove fallback in 1042, or remove all fallback here.

Reply:
I changed how I'm doing fallbacks. I think it looks cleaner now, but please take a look and let me know what you think.